src/console/CliManager.cpp

```diff
@@ -240,5 +243,313 @@ void CliManager::loadHistory() {
+        }
+        if (::strncasecmp(text.c_str(), word.c_str(), text.size()) == 0) {
+            matches.emplace_back(word);
+        }
+    }
+
+    return matches;
+}
+
+
+Status loadCompletions() {
+    using fs::FileUtils;
+    auto dir = FileUtils::readLink("/proc/self/exe").value();
+    dir = FileUtils::dirname(dir.c_str()) + "/../share/resources";
+    std::string file = dir + "/" + "completion.json";
+    auto status = Status::OK();
+    int fd = -1;
+    do {
+        fd = ::open(file.c_str(), O_RDONLY);
+        if (fd == -1) {
+            status = Status::Error("Failed to open `%s': %s",
+                                    file.c_str(), ::strerror(errno));
+            break;
+        }
+
+        auto len = ::lseek(fd, 0, SEEK_END);
+        if (len == 0) {
+            status = Status::Error("File `%s' is empty", file.c_str());
+            break;
+        }
+
+        auto buffer = std::make_unique<char[]>(len + 1);
+        ::lseek(fd, 0, SEEK_SET);
+        ::read(fd, buffer.get(), len);
+        buffer[len] = '\0';
+
+        std::string content;
+        content.assign(buffer.get(), len);
+
+        try {
+            status = parseKeywordsFromJson(folly::parseJson(content));
+        } catch (const std::exception &e) {
+            status = Status::Error("Illegal json `%s': %s", file.c_str(), e.what());
+            break;
+        }
+
+        if (!status.ok()) {
+            break;
+        }
+    } while (false);
+
+    if (fd != -1) {
+        ::close(fd);
+    }
+
+    return status;
+}
+
+
+Status parseKeywordsFromJson(const folly::dynamic &json) {
+    auto iter = json.find("keywords");
+    if (iter == json.items().end()) {
+        fprintf(stderr, "completions: no `keywords' found\n");
+        return Status::OK();
+    }
+
+    for (auto &pair : iter->second.items()) {
+        auto &pkw = pair.first;
+        primaryKeywords.emplace_back(pkw.asString());
+        auto subIter = pair.second.find("sub_keywords");
+        if (subIter == pair.second.items().end()) {
+            continue;
+        }
+        if (!subIter->second.isArray()) {
+            fprintf(stderr, "sub-keywords for `%s' should be an array\n",
+                    pkw.asString().c_str());
+            continue;
+        }
+        for (auto &subKey : subIter->second) {
+            if (!subKey.isString()) {
+                fprintf(stderr, "keyword name should be of type string\n");
+                break;
+            }
+            subKeywords[pkw.asString()].emplace_back(subKey.asString());
+        }
+    }
+
+    iter = json.find("typenames");
+    if (iter == json.items().end()) {
+        fprintf(stderr, "completions: no `typenames' found\n");
+        return Status::OK();
+    }
+
+    for (auto &tname : iter->second) {
+        typeNames.emplace_back(tname.asString());
+    }
+
+    return Status::OK();
+}
+
+
+void CliManager::initAutoCompletion() {
+    // The completion function
+    ::rl_attempted_completion_function = completer;
```

Comment:
Is there no default auto-complete module in readline?  This is a very common feature.

Reply:
Yes, the default one is to complete filenames.

Reply:
If you mean whether readline has builtin support to all these functions here(Given a collection of keywords, readline does all completion works for you). I'm afraid the answer seems no. I had been following up the official documents.

Reply:
。。。。The default module is useless if not allow users to pass custom keywords.

Reply:
Yes it is. Actually, you could infer from the `completer` function, that if we don't have to support sub-keywords completion, all we need is to retrieve the longest common prefix of all completion candidates.